### ISSUE TITLE
Missing typeerror for comparisons to null

### ISSUE THREAD

**Issue by KingSupernova31** [godotengine/godot#95825]
### Tested versions

4.3

### System information

Godot v4.3.stable - macOS 10.14.6 - GLES3 (Compatibility) - AMD Radeon R9 M370X OpenGL Engine - Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz (8 Threads)

### Issue description

```
var foo: String
if foo == 0:
	pass
```

This shows an error, `Invalid operands "String" and "int" for "==" operator.` But change the 0 to null, and the error will go away, despite it being just as invalid to compare a string to null.

### Steps to reproduce

N/A

### Minimal reproduction project (MRP)

N/A

**Comment by BenjaminCole1** [comment]
I think this is intentional. Since you did not give foo a value, it is null. 0 and null are not equal in GDScript as far as I am aware.

**Comment by KingSupernova31** [comment]
No, it's been statically-typed as a string, so it must have that type. Null is not a string, it's a separate variable type. See https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/static_typing.html.

Statically-typed variables that are initialized without a value are given that type's default value. The empty string for strings, false for booleans, 0 for ints and floats, etc.

**Comment by AThousandShips** [comment]
There's no operator between `String` and `int`, in either direction, all types have a `null` operator though, defined here:
https://github.com/godotengine/godot/blob/826de7976a6add282c7b14d4be2a7e6d775821d8/core/variant/variant_op.cpp#L571

**Comment by KingSupernova31** [comment]
Conceptually, the reason there's an error comparing two values of predefined types for equality is that it's a sign the programmer has made a mistake. `string_typed_variable == int_typed_variable` will _always_ return false, so there's never a reason someone would want to write that line of code in their program. Ergo, if that line exists, the programmer has done something wrong, and it's helpful to make them aware of that rather than to silently fail by always returning false for an expression that the programmer expects to sometimes return true.

The same consideration applies to null, even if it's not technically its own type. `any_typed_variable == null` will always return false, so that line existing in a program means the programmer has screwed up somewhere. So it should show an error.

I think this is especially important due to the existence of the error for other types. I didn't realize that null isn't technically a type until just now, and I'd guess that a large fraction of GDscript programmers don't realize it either, since it functions as a type in many practical ways. So if a programmer is used to the editor catching their variable typing mistakes when values _other_ than null are involved, they'll likely also expect it to catch an error like `string_variable == null`, and it will take them more debugging time to figure out what happened when their program doesn't work.

**Comment by AThousandShips** [comment]
This isn't a case of "this will always be false, you've made a mistake", that's not why the error happens, it's because there's no operator defined. That seems to be the cause of some confusion here

Making this an explicit check for pathological code is a separate issue and should be a proposal instead IMO

**Comment by KingSupernova31** [comment]
Why doesn't "foo" == 0 return an error?

**Comment by AThousandShips** [comment]
That would be something that should be an error IMO but not my area

**Comment by dalexeev** [comment]
The `==` operator has a bit confusing behavior. In some cases, comparing different types is allowed (and can return either `true` or `false`), in other cases it produces runtime errors. See:

* #26375
* #31331
* godotengine/godot-proposals#4571
* godotengine/godot-proposals#5101
* godotengine/godot-proposals#6092

The static analyzer tries to match the runtime behavior, and produces an error only if it knows in advance that this will cause an error at runtime.

Unlike the `==` operator, the `is_same()` function never produces errors, it just returns `false`. Note that `is_same()` uses a stricter comparison.

See also [GDScript reference - Operators](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#operators):

> The == and != operators sometimes allow you to compare values of different types (for example, 1 == 1.0 is true), but in other cases it can cause a runtime error. If you're not sure about the types of the operands, you can safely use the is_same() function (but note that it is more strict about types and references). To compare floats, use the is_equal_approx() and is_zero_approx() functions instead.

**Comment by KingSupernova31** [comment]
Oh, I was mistaken above; `"foo" == 0` does return an error, as does `untyped_variable_that_is_currently_a_string == 0`. That is in my opinion very poor design, since it defeats the purpose of having untyped variables available in the language in the first place.

But that's not the same issue I'm bringing up here. I'm saying that null functions as a type in most practical ways and therefore it should be treated as one for the purpose of informing the programmer that their code contains a problem. If the editor is going to warn or error for mixing up types, it should also warn or error for mixing up a type with null.
